### PR TITLE
[Core][Distributed] Refactor ipc buffer init in CustomAllreduce

### PR DESCRIPTION
**Pull request by hanzhi713**
As discussed with @youkaichao, we use cuda API to share tensors instead of replying `_share_cuda_`, which won't break with expandable segment or future pytorch upgrade.

Additional changes:
1. Improved some comments, especially about cuda graph ipc registration.
2. Consolidated `all_reduce` methods. 
3. Use `vector<int64_t>` instead of pytorch tensor for handle during cuda graph ipc registration process. The use of Tensor was introduced in #5047. We sticked to non-tensor type to remove this complexity.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by mgoin** [comment]
cc @tlrmchlsmth

**Review by tlrmchlsmth — COMMENTED** [review]
@hanzhi713 and @youkaichao could you share a few more details on what's going on? 

Looks like this is related to https://github.com/vllm-project/vllm/pull/9815 -- is the idea to be more stable across Pytorch versions? Do you see any downsides to this?

**Comment by hanzhi713** [comment]
> @hanzhi713 and @youkaichao could you share a few more details on what's going on?
> 
> Looks like this is related to #9815 -- is the idea to be more stable across Pytorch versions? Do you see any downsides to this?

Yes, we want to rely less on internal API to prevent future breaking. A downside to the current approach is that it doesn't support `expandable_segments:True`. 

An alternative design I see is to do the one-time allocation of IPC-enabled buffers ourselves through CUDA C++ (i.e. cudaMalloc + ipc handle calls).

**Comment by youkaichao** [comment]
https://github.com/vllm-project/vllm/pull/10064 will make this pr easier. we don't need to depend on pytorch's internal apis

**Comment by youkaichao** [comment]
please merge main to use the functionality from https://github.com/vllm-project/vllm/pull/10064

**Review by youkaichao — APPROVED** [review]
thanks for the great contribution! do you have some updated perf numbers? I assume it should not affect the performance.

**Comment by youkaichao** [comment]
nit: please fix the format.

**Comment by hanzhi713** [comment]
There're no perf changes from the C++. I can run the python side benchmark to be sure.

**Comment by hanzhi713** [comment]
@youkaichao I can confirm that there's no perf difference with `benchmarks/benchmark_latency.py`

**Comment by youkaichao** [comment]
> RuntimeError: Inplace update to inference tensor outside InferenceMode is not allowed.You can make a clone to get a normal tensor before doing inplace

there are some errors in the ci actually @hanzhi713

File: vllm/distributed/device_communicators/custom_all_reduce.py [hunk 1]
```diff
@@ -220,60 +219,14 @@ def capture(self):
...
-    def register_buffer(self, inp: torch.Tensor):
-        handles, offsets = self._get_ipc_meta(inp)
-        ops.register_buffer(self._ptr, inp, handles, offsets)
-
     def register_graph_buffers(self):
         handle, offset = ops.get_graph_buffer_ipc_meta(self._ptr)
-        handles, offsets = self._gather_ipc_meta((bytes(handle), offset))
         logger.info("Registering %d cuda graph addresses", len(offset))
+        all_data = [None] * dist.get_world_size(group=self.group)
+        dist.all_gather_object(all_data, (handle, offset), group=self.group)
```
Comment (youkaichao) [inline vllm/distributed/device_communicators/custom_all_reduce.py:226]: use broadcast as before in `_gather_ipc_meta` ?

code for reference:

```python
        for i, rank in enumerate(ranks):
            dist.broadcast_object_list(all_data[i],
                                       src=rank,
                                       group=self.group,
                                       device="cpu")
```

Comment (hanzhi713) [reply vllm/distributed/device_communicators/custom_all_reduce.py:226]: Yeah it looks like we still need to use broadcast here

Comment (youkaichao) [reply vllm/distributed/device_communicators/custom_all_reduce.py:226]: to locally test it, run `pytest -v -s tests/basic_correctness/test_basic_correctness.py::test_models_distributed[facebook/opt-125m-mp--A100]`

Comment (hanzhi713) [reply vllm/distributed/device_communicators/custom_all_reduce.py:226]: My old benchmark script had `--enforce-eager there` which didn't catch this

Comment (hanzhi713) [reply vllm/distributed/device_communicators/custom_all_reduce.py]: Should be fixed now.

**Comment by youkaichao** [comment]
@hanzhi713 thanks again for the great contribution!

**Pull request merged**